### PR TITLE
Migrate elastic_agent integration to package spec v3

### PR DESCRIPTION
**Pull request by belimawr**
## Proposed commit message

Migrate the elastic_agent integration to package spec v3.
 - Mappings were updated as required
 - Missing filters (SVR00002) errors are now ignored because some dashboards cannot have a common filter in all visualistaions.
 
Changes to dashboards:
- [Elastic Agent] Agent Info
  - Removed missing reference

- [Elastic Agent] Agent metrics
  - Visualisations were migrated to lens
  - cgroup CPU percentage was removed from the Elastic-Agent Metrics dashboard because we don't collect those metrics on a default installation with monitoring enabled

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

~~## Author's Checklist~~

## How to test this PR locally

1. Using `elastic-package` build the elastic_agent integration
2. Using `elastic-package` bring up the stack
3. Go to the dashboards:
    - [Elastic Agent] Agent Info
    - [Elastic Agent] Integrations
    - [Elastic Agent] Agent metrics

## Related issues

- Closes https://github.com/elastic/integrations/issues/10275

## Screenshots
Comparison between old and new CPU usage visualisation in the Elastic-Agent dashboard. The CPU usage is now aggregated by `component.id`, effectively aggregating by running process, while the old one aggregated by binary name.

At the top is the old visualisation and the new one is at the bottom.

### CODE REVIEW

**Comment by jlind23** [comment]
@belimawr updating the description in order to link - https://github.com/elastic/integrations/issues/10275

File: packages/elastic_agent/data_stream/endpoint_security_metrics/fields/fields.yml [hunk 1]
```diff
@@ -183,6 +183,7 @@
               description: CPU time consumed by tasks in user (kernel) mode.
             - name: percpu
               type: object
+              object_type: object
```
Comment (jsoriano) [inline packages/elastic_agent/data_stream/endpoint_security_metrics/fields/fields.yml:186]:
```suggestion
              object_type: long
```

File: packages/elastic_agent/data_stream/cloud_defend_logs/fields/fields.yml [hunk 2]
```diff
@@ -2,8 +2,10 @@
   type: text
 - name: input
   type: object
+  object_type: object
```
Comment (jsoriano) [inline packages/elastic_agent/data_stream/cloud_defend_logs/fields/fields.yml:5]: I guess we only need `input.type`.
```suggestion
- name: input.type
  description: "Input type"
  type: keyword
```

File: packages/elastic_agent/data_stream/cloud_defend_logs/fields/fields.yml [hunk 3]
```diff
@@ -2,8 +2,10 @@
...
 - name: result
   type: object
+  object_type: object
```
Comment (jsoriano) [inline packages/elastic_agent/data_stream/cloud_defend_logs/fields/fields.yml:8]: Is this field needed? I could not find a document with this field.
```suggestion
```

File: packages/elastic_agent/data_stream/cloudbeat_logs/fields/fields.yml [hunk 4]
```diff
@@ -2,8 +2,10 @@
   type: text
 - name: input
   type: object
+  object_type: object
```
Comment (jsoriano) [inline packages/elastic_agent/data_stream/cloudbeat_logs/fields/fields.yml:5]:
```suggestion
- name: input.type
  description: "Input type"
  type: keyword
```

**Review by jsoriano — COMMENTED** [review]
I think there is some issue with `object_type: object`, I would try to avoid it when possible, adding some suggestions for that.

File: packages/elastic_agent/data_stream/cloud_defend_logs/fields/fields.yml (hunk 3 above)
Comment (belimawr) [reply packages/elastic_agent/data_stream/cloud_defend_logs/fields/fields.yml:8]: I don't know. I've never run Clouddefend, I'm on a best-effort approach here. :sweat_smile:

File: packages/elastic_agent/kibana/dashboard/elastic_agent-1a4e7280-6b5e-11ed-98de-67bdecd21824.json [hunk 5]
```diff
@@ -529,5 +567,5 @@
         }
     ],
     "type": "dashboard",
-    "typeMigrationVersion": "8.9.0"
+    "typeMigrationVersion": "10.2.0"
```
Comment (jsoriano) [inline packages/elastic_agent/kibana/dashboard/elastic_agent-1a4e7280-6b5e-11ed-98de-67bdecd21824.json:570]: Where does this version comes from? :thinking:

Comment (belimawr) [reply packages/elastic_agent/kibana/dashboard/elastic_agent-1a4e7280-6b5e-11ed-98de-67bdecd21824.json:570]: Honestly, I don't know... I just edited the dashboards on Kibana and then exported them using `elastic-package export dashboards`.

I was testing again and it seems I edited this one by mistake, I'll revert all changes on this dashboard.

File: packages/elastic_agent/kibana/dashboard/elastic_agent-0600ffa0-6b5e-11ed-98de-67bdecd21824.json [hunk 6]
```diff
@@ -864,5 +902,5 @@
         }
     ],
     "type": "dashboard",
-    "typeMigrationVersion": "8.9.0"
+    "typeMigrationVersion": "10.2.0"
```
Comment (jsoriano) [inline packages/elastic_agent/kibana/dashboard/elastic_agent-0600ffa0-6b5e-11ed-98de-67bdecd21824.json:905]: There is still a reference to this version here.

File: packages/elastic_agent/kibana/dashboard/elastic_agent-f47f18cc-9c7d-4278-b2ea-a6dee816d395.json [hunk 7]
```diff
@@ -3274,12 +3292,17 @@
...
+        },
         {
             "id": "metrics-*",
             "name": "controlGroup_2678bf39-3def-453e-9f30-2904bc88efe9:optionsListDataView",
             "type": "index-pattern"
         }
     ],
     "type": "dashboard",
-    "typeMigrationVersion": "8.9.0"
+    "typeMigrationVersion": "10.2.0"
```
Comment (jsoriano) [inline packages/elastic_agent/kibana/dashboard/elastic_agent-f47f18cc-9c7d-4278-b2ea-a6dee816d395.json:3307]: And here.

File: packages/elastic_agent/kibana/dashboard/elastic_agent-0600ffa0-6b5e-11ed-98de-67bdecd21824.json (hunk 6 above)
Comment (jlind23) [inline packages/elastic_agent/kibana/dashboard/elastic_agent-0600ffa0-6b5e-11ed-98de-67bdecd21824.json:905]:
```suggestion
    "typeMigrationVersion": "8.9.0"
```

File: packages/elastic_agent/kibana/dashboard/elastic_agent-f47f18cc-9c7d-4278-b2ea-a6dee816d395.json (hunk 7 above)
Comment (jlind23) [inline packages/elastic_agent/kibana/dashboard/elastic_agent-f47f18cc-9c7d-4278-b2ea-a6dee816d395.json:3307]:
```suggestion
    "typeMigrationVersion": "8.9.0"
```

**Review by jlind23 — APPROVED** [review]

**Comment by jlind23** [comment]
Agent metrics is working but the agent info one fails with `Could not find reference "panel_9604578e-7da2-4575-923e-f15e51bca436"`
Screenshot 2024-07-10 at 13 36 37

**Comment by jlind23** [comment]
Both dashboards are working now, I have been able to test them locally.

File: packages/elastic_agent/changelog.yml [hunk 8]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "2.0.0"
+  changes:
+    - description: Update package spec to 3.1.4
+      type: enhancement
+      link: "foo"
```
Comment (carsonip) [inline packages/elastic_agent/changelog.yml:6]: changelog link missing

**Comment by jlind23** [comment]
@belimawr It looks like along the way your removed:
`        {
            "id": "elastic_agent-522c9e20-ad53-11ed-957f-f1c897630287",
            "name": "9604578e-7da2-4575-923e-f15e51bca436:panel_9604578e-7da2-4575-923e-f15e51bca436",
            "type": "search"
        },`

Which gave the ability to show Agent Errors, what was the reason why you removed it? If there was no reason and just a mistake then we should most probably revert this change.
cc @anmironov @pierrehilbert

**Comment by belimawr** [comment]
> @belimawr It looks like along the way your removed: ` { "id": "elastic_agent-522c9e20-ad53-11ed-957f-f1c897630287", "name": "9604578e-7da2-4575-923e-f15e51bca436:panel_9604578e-7da2-4575-923e-f15e51bca436", "type": "search" },`
> 
> Which gave the ability to show Agent Errors, what was the reason why you removed it? If there was no reason and just a mistake then we should most probably revert this change. cc @anmironov @pierrehilbert

Looking back at the commit that removed I believe it was removed because of this lint message:
```
Error: linting package failed: found 1 validation error:
   1. references found in dashboard kibana/dashboard/elastic_agent-0600ffa0-6b5e-11ed-98de-67bdecd21824.json: elastic_agent-522c9e20-ad53-11ed-957f-f1c897630287 (search) (SVR00004)
```

I might have "fixed" it the wrong way :man_facepalming:.

I tried reverting the change, and I get the lint error again. My integrations knowledge is limited, so I'm not sure the correct way to fix it.

**Comment by jsoriano** [comment]
> I tried reverting the change, and I get the lint error again. My integrations knowledge is limited, so I'm not sure the correct way to fix it.

This problem indicates that there is something in a dashboard that is not referenced by value, what is required now for most objects.

There are two options here to fix this, you can edit the dashboard and embed the search in the dashboard, or if this is not possible now, to add the `SVR00004` code to the list of excluded checks in `packages/elastic_agent/validation.yml`.

**Comment by belimawr** [comment]
Thanks @jsoriano!

I tried inlining the searches using https://github.com/elastic/visualizations_integrations_tools, but it did not work, so I added the exception to the lint, here is the PR: https://github.com/elastic/integrations/pull/10728

**Pull request merged**